test/libyul/ObjectCompilerTest.cpp

```diff
@@ -0,0 +1,124 @@
+
+#include <test/libsolidity/FormattedScope.h>
+
+#include <libsolidity/interface/AssemblyStack.h>
+
+#include <libevmasm/Instruction.h>
+
+#include <liblangutil/SourceReferenceFormatter.h>
+
+#include <boost/algorithm/string.hpp>
+
+using namespace dev;
+using namespace langutil;
+using namespace yul;
+using namespace yul::test;
+using namespace dev::solidity;
+using namespace dev::solidity::test;
+using namespace std;
+
+ObjectCompilerTest::ObjectCompilerTest(string const& _filename)
+{
+	boost::filesystem::path path(_filename);
+
+	ifstream file(_filename);
+	if (!file)
+		BOOST_THROW_EXCEPTION(runtime_error("Cannot open test case: \"" + _filename + "\"."));
+	file.exceptions(ios::badbit);
+
+	string line;
+	while (getline(file, line))
+	{
+		if (boost::algorithm::starts_with(line, "// ----"))
+			break;
+		if (m_source.empty() && boost::algorithm::starts_with(line, "// optimize"))
+			m_optimize = true;
+		m_source += line + "\n";
+	}
+	while (getline(file, line))
+		if (boost::algorithm::starts_with(line, "// "))
+			m_expectation += line.substr(3) + "\n";
+		else
+			m_expectation += line + "\n";
+}
+
+bool ObjectCompilerTest::run(ostream& _stream, string const& _linePrefix, bool const _formatted)
+{
+	AssemblyStack stack(EVMVersion(), AssemblyStack::Language::StrictAssembly);
+	if (!stack.parseAndAnalyze("source", m_source))
+	{
+		FormattedScope(_stream, _formatted, {formatting::BOLD, formatting::RED}) << _linePrefix << "Error parsing source." << endl;
+		printErrors(_stream, stack.errors(), stack.scanner());
+		return false;
+	}
+	if (m_optimize)
+		stack.optimize();
+	MachineAssemblyObject obj = stack.assemble(AssemblyStack::Machine::EVM);
+	solAssert(obj.bytecode, "");
+	m_obtainedResult =
+		"Assembly:\n" +
+		obj.assembly +
+		"Bytecode: " +
+		toHex(obj.bytecode->bytecode) +
+		"\nOpcodes: " +
+		boost::trim_copy(solidity::disassemble(obj.bytecode->bytecode)) +
+		"\n";
+
+	if (m_expectation != m_obtainedResult)
+	{
+		string nextIndentLevel = _linePrefix + "  ";
+		FormattedScope(_stream, _formatted, {formatting::BOLD, formatting::CYAN}) << _linePrefix << "Expected result:" << endl;
+		printIndented(_stream, m_expectation, nextIndentLevel);
+		FormattedScope(_stream, _formatted, {formatting::BOLD, formatting::CYAN}) << _linePrefix << "Obtained result:" << endl;
+		printIndented(_stream, m_obtainedResult, nextIndentLevel);
+		return false;
+	}
+	return true;
+}
+
+void ObjectCompilerTest::printSource(ostream& _stream, string const& _linePrefix, bool const) const
+{
+	printIndented(_stream, m_source, _linePrefix);
+}
+
+void ObjectCompilerTest::printUpdatedExpectations(ostream& _stream, string const& _linePrefix) const
+{
+	printIndented(_stream, m_obtainedResult, _linePrefix);
+}
+
+void ObjectCompilerTest::printIndented(ostream& _stream, string const& _output, string const& _linePrefix) const
+{
+	stringstream output(_output);
+	string line;
+	while (getline(output, line))
+		_stream << _linePrefix << line << endl;
```

Comment:
I haven't really looked into it, but to avoid trailing whitespace for empty lines this could probably just be replaced by something like
```
if (line.empty()) _stream << trim(_linePrefix) << endl;
else _stream << _linePrefix << line << endl;
```
?

Reply:
Yeah, would be nice if we would not have to do that in all tests...

Reply:
Or actually ``boost::trim_right_copy`` or something?

Reply:
Yeah, would be nicer to always do it in isoltest - but that would need some refactoring. Actually not too hard, though - I could look into it?

Reply:
I tried it, it results in "failure" after you run 'update'.

The current version seems to work, so I won't waste another hour on it for now :)

Reply:
Ok!